bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManagerImpl.java

```diff
@@ -1111,6 +1128,61 @@ private void setThingStatus(Thing thing, ThingStatusInfo thingStatusInfo) {
         }
     }
 
+    @Override
+    public void enableThing(ThingUID thingUID, boolean enabled) {
+        Thing thing = getThing(thingUID);
+        ThingHandler thingHandler = thing.getHandler();
+        persistEnabledStatusForThing(thing.getUID(), enabled);
+
+        if (enabled) {
+            logger.debug("Thing {} will be enabled.", thing.getUID().getAsString());
+            if (thingHandler != null && isHandlerRegistered(thing)) {
+                initializeHandler(thing);
+            } else {
+                registerAndInitializeHandler(thing, findThingHandlerFactory(thing.getThingTypeUID()));
+            }
+        } else {
+            logger.debug("Thing {} will be disabled.", thing.getUID().getAsString());
+            if (thingHandler != null) {
+                unregisterAndDisposeHandler(findThingHandlerFactory(thing.getThingTypeUID()), thing, thingHandler);
+            }
+        }
+
+    }
+
+    /**
+     * Persists whether the thing should be enabled or disabled.
+     * 
+     * @param thingUID the uid of the thing. Will be used as a key.
+     * @param enabled
+     * <p>
+     * <ul>
+     * <li>case {@code true} the record for that key will be removed from the storage. Thing should
+     * be considered as enabled.
+     * <li>case {@code false} a boolean value false with a key the UID of the thing will be
+     * persisted. Thing should be considered as explicitly disabled.
+     * </ul>
+     * <p>
+     */
+    private void persistEnabledStatusForThing(final ThingUID thingUID, boolean enabled) {
+        // Persist only disabled things. Thing with no corresponding persisted value will be considered as enabled.
+        if (enabled) {
+            storage.remove(thingUID.getAsString());
+        } else {
+            storage.put(thingUID.getAsString(), new ManagedThingStatusInfo(enabled));
+        }
+    }
+
+    private boolean isPersistedAsEnabled(ThingUID thingUID) {
+        ManagedThingStatusInfo persistedStatus = storage.get(thingUID.getAsString());
```

Comment:
This method should simply be based on the ThingStatusDetail, see the [rule engine as an example](https://github.com/kaikreuzer/smarthome/blob/circular_dependencies/bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineImpl.java#L910-L913).
You should imho add a check in the `thingAdded` method, whether there is a disabled flag persisted in the storage and set the Thing status there already (and avoid a call of `registerAndInitializeHandler` at the same time.